File: smoke_tests/lua/race_condition.lua
-- Race Condition vulnerabilities in Lua

-- Test 1: TOCTOU in file operations
function safe_delete(path)
    -- VULNERABLE: TOCTOU race condition
    if file_exists(path) then
        -- Time gap between check and use
        os.remove(path)
    end
end

-- Test 2: Balance check race condition
function withdraw(user_id, amount)
    local balance = get_balance(user_id)
    -- VULNERABLE: Balance can change between check and update
    if balance >= amount then
        update_balance(user_id, balance - amount)
        return true
    end
    return false
end

-- Test 3: Inventory race condition
function purchase_item(user_id, item_id, quantity)
    local stock = get_stock(item_id)
    -- VULNERABLE: Stock can change
    if stock >= quantity then
        decrease_stock(item_id, quantity)
        add_to_cart(user_id, item_id, quantity)
    end
end

-- Test 4: Coupon usage race
function apply_coupon(coupon_code, order_id)
    local coupon = get_coupon(coupon_code)
    -- VULNERABLE: Coupon can be used multiple times
    if coupon and coupon.uses_remaining > 0 then
        apply_discount(order_id, coupon.discount)
        decrement_coupon_uses(coupon_code)
    end
end

-- Test 5: Session race condition
function login(username, password, session)
    if verify_credentials(username, password) then
        -- VULNERABLE: Session manipulation during login
        session.user = username
        session.logged_in = true
        session.login_time = os.time()
    end
end

-- Test 6: File creation race
function create_temp_file(data)
    local path = "/tmp/app_" .. os.time()
    -- VULNERABLE: File can be created by attacker
    if not file_exists(path) then
        local file = io.open(path, "w")
        file:write(data)
        file:close()
    end
end

-- Test 7: Lock-free counter
local request_count = 0

function handle_request()
    -- VULNERABLE: Non-atomic increment
    request_count = request_count + 1
    -- Process request...
end

-- Test 8: Rate limiting race
function check_rate_limit(user_id)
    local count = get_request_count(user_id)
    -- VULNERABLE: Race in rate limiting
    if count < 100 then
        increment_request_count(user_id)
        return true
    end
    return false
end

-- Test 9: Double spending
function process_payment(user_id, amount)
    local pending = get_pending_transactions(user_id)
    -- VULNERABLE: Can submit multiple transactions
    if not pending then
        set_pending(user_id, true)
        process_transaction(user_id, amount)
        set_pending(user_id, false)
    end
end

-- Test 10: Sequence number race
local sequence = 0

function get_next_sequence()
    -- VULNERABLE: Non-atomic operation
    sequence = sequence + 1
    return sequence
end

-- Test 11: Cache race condition
function get_cached_data(key)
    local data = cache:get(key)
    -- VULNERABLE: Cache can be modified during processing
    if not data then
        data = fetch_from_db(key)
        cache:set(key, data)
    end
    return data
end

-- Test 12: User registration race
function register_user(username, password)
    -- VULNERABLE: Username can be taken between check and insert
    if not user_exists(username) then
        create_user(username, password)
        return true
    end
    return false
end

-- Test 13: Shared state in coroutines
local shared_data = {}

function coroutine_handler(key, value)
    -- VULNERABLE: Shared state between coroutines
    shared_data[key] = value
    coroutine.yield()
    -- Value may have changed
    return shared_data[key]
end

-- Helper functions (stubs)
function file_exists(path) return false end
function get_balance(id) return 100 end
function update_balance(id, amount) end
function get_stock(id) return 10 end
function decrease_stock(id, qty) end
function add_to_cart(uid, iid, qty) end
function get_coupon(code) return {uses_remaining = 1, discount = 10} end
function apply_discount(order, discount) end
function decrement_coupon_uses(code) end
function verify_credentials(u, p) return true end
function get_request_count(id) return 0 end
function increment_request_count(id) end
function get_pending_transactions(id) return nil end
function set_pending(id, val) end
function process_transaction(id, amt) end
function fetch_from_db(key) return {} end
function user_exists(username) return false end
function create_user(username, password) end
cache = { get = function(k) return nil end, set = function(k, v) end }

File: smoke_tests/php/race_condition.php
<?php
// Race Condition vulnerabilities in PHP

// Test 1: Check-then-act on balance
function withdraw() {
    $amount = (int)$_POST['amount'];
    $user_id = $_SESSION['user_id'];

    $balance = get_balance($user_id);
    // VULNERABLE: Race between check and update
    if ($balance >= $amount) {
        usleep(10000);  // Simulates processing
        update_balance($user_id, $balance - $amount);
        echo json_encode(['new_balance' => $balance - $amount]);
    } else {
        echo json_encode(['error' => 'Insufficient funds']);
    }
}

// Test 2: File TOCTOU
function read_config() {
    $filename = $_GET['filename'];
    $path = "/config/$filename";

    // VULNERABLE: File can change between check and read
    if (file_exists($path)) {
        usleep(10000);
        $content = file_get_contents($path);
        echo $content;
    }
}

// Test 3: Coupon redemption race
function redeem_coupon() {
    $coupon_code = $_POST['code'];

    $coupon = get_coupon($coupon_code);
    // VULNERABLE: Race to redeem same coupon multiple times
    if ($coupon && !$coupon['used']) {
        apply_discount($coupon);
        mark_coupon_used($coupon_code);
        echo "Discount applied!";
    }
}

// Test 4: Session race condition
function update_cart() {
    session_start();
    // VULNERABLE: Session not locked during operation
    $cart = $_SESSION['cart'] ?? [];
    usleep(10000);
    $cart[] = $_POST['item'];
    $_SESSION['cart'] = $cart;
}

// Test 5: File write race
function write_log() {
    $content = $_POST['content'];
    $path = '/var/log/app.log';

    // VULNERABLE: Concurrent writes can interleave
    $f = fopen($path, 'a');
    fwrite($f, $content);
    fclose($f);
}

// Test 6: Counter increment race
function increment_counter() {
    $counter_file = '/tmp/counter.txt';

    // VULNERABLE: Non-atomic increment
    $count = (int)file_get_contents($counter_file);
    usleep(10000);
    file_put_contents($counter_file, $count + 1);
    echo $count + 1;
}

// Test 7: User registration race
function register_user() {
    $email = $_POST['email'];

    // VULNERABLE: Race between check and insert
    if (!user_exists($email)) {
        usleep(10000);
        create_user($email, $_POST['password']);
        echo "User created";
    }
}

// Test 8: File creation race
function create_unique_file() {
    $filename = $_GET['name'];
    $path = "/uploads/$filename";

    // VULNERABLE: TOCTOU race
    if (!file_exists($path)) {
        usleep(10000);
        file_put_contents($path, $_POST['content']);
    }
}

// Test 9: Lock file race
function acquire_lock() {
    $lock_file = '/tmp/app.lock';

    // VULNERABLE: Non-atomic lock check
    if (!file_exists($lock_file)) {
        touch($lock_file);
        // Do work
        unlink($lock_file);
    }
}

// Test 10: Database race (SELECT then UPDATE)
function update_stock() {
    $product_id = $_POST['product_id'];
    $quantity = (int)$_POST['quantity'];

    $pdo = get_pdo();
    // VULNERABLE: Race between SELECT and UPDATE
    $stmt = $pdo->prepare("SELECT stock FROM products WHERE id = ?");
    $stmt->execute([$product_id]);
    $stock = $stmt->fetchColumn();

    if ($stock >= $quantity) {
        usleep(10000);
        $stmt = $pdo->prepare("UPDATE products SET stock = stock - ? WHERE id = ?");
        $stmt->execute([$quantity, $product_id]);
    }
}

// Test 11: Directory creation race
function ensure_directory() {
    $dir = $_GET['dir'];
    $path = "/data/$dir";

    // VULNERABLE: Race condition
    if (!is_dir($path)) {
        mkdir($path, 0755, true);
    }
}

// Helper functions
function get_balance($user_id) { return 1000; }
function update_balance($user_id, $balance) {}
function get_coupon($code) { return ['used' => false]; }
function apply_discount($coupon) {}
function mark_coupon_used($code) {}
function user_exists($email) { return false; }
function create_user($email, $password) {}
function get_pdo() { return new PDO('sqlite::memory:'); }
?>
